src/FixedFrequencyAgent.cpp

```diff
@@ -0,0 +1,309 @@
+        }
+    }
+
+    // Indicate whether to send the policy down to children
+    bool FixedFrequencyAgent::do_send_policy(void) const
+    {
+        return true;
+    }
+
+    void FixedFrequencyAgent::aggregate_sample(const std::vector<std::vector<double> > &in_sample,
+                                        std::vector<double>& out_sample)
+    {
+
+    }
+
+    // Indicate whether to send samples up to the parent
+    bool FixedFrequencyAgent::do_send_sample(void) const
+    {
+        return false;
+    }
+
+    void FixedFrequencyAgent::adjust_platform(const std::vector<double>& in_policy)
+    {
+        assert(in_policy.size() == M_NUM_POLICY);
+        m_do_write_batch = false;
+
+        double accel_freq_request = in_policy[M_POLICY_ACCELERATOR_FREQUENCY];
+        double cpu_freq_request = in_policy[M_POLICY_CPU_FREQUENCY];
+        double uncore_freq_request = in_policy[M_POLICY_UNCORE_FREQUENCY];
+
+        // set Accelerator frequency control
+        auto freq_ctl_itr = m_control_available.find("FREQUENCY_ACCELERATOR_CONTROL");
+        if (!std::isnan(in_policy[M_POLICY_ACCELERATOR_FREQUENCY])) {
+            if (accel_freq_request != freq_ctl_itr->second.m_last_setting) {
+                m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, accel_freq_request);
+                freq_ctl_itr->second.m_last_setting = accel_freq_request;
+                m_do_write_batch = true;
+            }
+        }
+
+        // set CPU frequency control
+        freq_ctl_itr = m_control_available.find("CPU_FREQUENCY_CONTROL");
+        if (!std::isnan(in_policy[M_POLICY_CPU_FREQUENCY])) {
+            if (cpu_freq_request != freq_ctl_itr->second.m_last_setting) {
+                m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, cpu_freq_request);
+                freq_ctl_itr->second.m_last_setting = cpu_freq_request;
+                m_do_write_batch = true;
+            }
+        }
+
+        // set Uncore frequency control
+        freq_ctl_itr = m_control_available.find("MSR::UNCORE_RATIO_LIMIT:MIN_RATIO");
+        if (!std::isnan(in_policy[M_POLICY_UNCORE_FREQUENCY])) {
+            if (uncore_freq_request != freq_ctl_itr->second.m_last_setting) {
+                m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, uncore_freq_request);
+                freq_ctl_itr->second.m_last_setting = uncore_freq_request;
+                m_do_write_batch = true;
+            }
+            freq_ctl_itr = m_control_available.find("MSR::UNCORE_RATIO_LIMIT:MAX_RATIO");
+            if (uncore_freq_request != freq_ctl_itr->second.m_last_setting) {
+                m_platform_io.adjust(freq_ctl_itr->second.m_batch_idx, uncore_freq_request);
+                freq_ctl_itr->second.m_last_setting = uncore_freq_request;
+                m_do_write_batch = true;
+            }
+        }
+
+        if (!m_is_adjust_initialized) {
+            double cpu_init_freq = m_platform_io.read_signal("CPU_FREQUENCY_CONTROL", GEOPM_DOMAIN_BOARD, 0);
```

Comment:
We're handling the initial CPU & Uncore frequency control here, but the GPU control above (through guaranteeing a policy value).  It may be better to just explicitly handle all initial writes here.

Reply:
Fixed